actors/builtin/miner/miner_actor.go

```diff
@@ -2067,6 +2068,253 @@ func (a Actor) RepayDebt(rt Runtime, _ *abi.EmptyValue) *abi.EmptyValue {
+			newSectorInfo.ExpectedStoragePledge = ExpectedRewardForPower(rewRet.ThisEpochRewardSmoothed, powRet.QualityAdjPowerSmoothed, pwr, InitialPledgeProjectionPeriod)
+			newSectorInfo.ReplacedSectorAge = maxEpoch(0, rt.CurrEpoch()-updateWithSectorInfo.sectorInfo.Activation)
+
+			initialPledgeAtUpgrade := InitialPledgeForPower(pwr, rewRet.ThisEpochBaselinePower, rewRet.ThisEpochRewardSmoothed,
+				powRet.QualityAdjPowerSmoothed, rt.TotalFilCircSupply())
+
+			if initialPledgeAtUpgrade.GreaterThan(updateWithSectorInfo.sectorInfo.InitialPledge) {
+				deficit := big.Sub(initialPledgeAtUpgrade, updateWithSectorInfo.sectorInfo.InitialPledge)
+
+				unlockedBalance, err := st.GetUnlockedBalance(rt.CurrentBalance())
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to calculate unlocked balance")
+				if unlockedBalance.LessThan(deficit) {
+					rt.Log(rtt.INFO, "insufficient funds for new initial pledge requirement %s, available: %s, skipping sector %d",
+						deficit, unlockedBalance, updateWithSectorInfo.sectorInfo.SectorNumber)
+					continue
+				}
+
+				err = st.AddInitialPledge(deficit)
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to add initial pledge")
+
+				newSectorInfo.InitialPledge = initialPledgeAtUpgrade
+			}
+
+			deadlines, err := st.LoadDeadlines(store)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
+
+			deadline, err := deadlines.LoadDeadline(store, updateWithSectorInfo.update.Deadline)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadline %d", updateWithSectorInfo.update.Deadline)
+
+			partitions, err := deadline.PartitionsArray(store)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load partitions for deadline %d", updateWithSectorInfo.update.Deadline)
+
+			quant := st.QuantSpecForDeadline(updateWithSectorInfo.update.Deadline)
+
+			var partition Partition
+			found, err := partitions.Get(updateWithSectorInfo.update.Partition, &partition)
+
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadline %v partition %v",
+				updateWithSectorInfo.update.Deadline, updateWithSectorInfo.update.Partition)
+
+			if !found {
+				rt.Abortf(exitcode.ErrNotFound, "no such deadline %v partition %v", updateWithSectorInfo.update.Deadline, updateWithSectorInfo.update.Partition)
+			}
+
+			// TODO: Group by partition
+			partitionPowerDelta, partitionPledgeDelta, err := partition.ReplaceSectors(store,
+				[]*SectorOnChainInfo{updateWithSectorInfo.sectorInfo},
+				[]*SectorOnChainInfo{newSectorInfo},
+				info.SectorSize,
+				quant)
+
+			// TODO: actual error / continue
+			if err != nil {
+				return
+			}
+
+			powerDelta = powerDelta.Add(partitionPowerDelta)
+			pledgeDelta = big.Add(pledgeDelta, partitionPledgeDelta) // expected to be zero, see note below.
+
+			err = partitions.Set(updateWithSectorInfo.update.Partition, &partition)
```

Comment:
Right now this isn't persisted you need to flush partition and deadline (`partitions.Root()` => `deadline.UpdateDeadline` => `SaveDeadlines`)

Reply:
Check out `ExtendSectorExpiration`

Reply:
lmao nice, i'm an amoeba brain